### PR TITLE
Behat tests updates

### PR DESCRIPTION
**Pull request by dalailomo**

### CODE REVIEW

File: hub/api/dic.php [hunk 1]
```diff
@@ -2,7 +2,8 @@
 
 $app['plugins'] = $app->share(function ($app) {
     $pluginsCollection = new \FP\Larmo\Domain\Service\PluginsCollection;
-    $pluginsRepository = new \FP\Larmo\Infrastructure\Repository\FilesystemPlugins($app['config.path.plugins']);
+    $directoryIterator = new \DirectoryIterator($app['config.path.plugins']);
```
Comment (zeulus) [inline hub/api/dic.php]: Why did you change that behavior?

File: hub/tests/features/HandleWellFormedPacket.feature [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Feature: Handle a well formed packet coming from an agent
+  Background:
+    Given A well formed packet coming from an agent
+
+  Scenario: Plugin service successes to validate
+    When The packet is prepared to be verified against a schema
```
Comment (zeulus) [inline hub/tests/features/HandleWellFormedPacket.feature]: schema validation and plugin validation are two separate things - they shouldn't come together in one story

File: hub/tests/features/HandleWellFormedPacket.feature [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    Then The plugin service successes to validate
+
+  Scenario: A packet can be created by the system
```
Comment (zeulus) [inline hub/tests/features/HandleWellFormedPacket.feature]: that scenarion doesn't tell the story - you're creating packet object from array, that's what it's testing, nothing else

File: hub/tests/features/HandleWellFormedPacket.feature [hunk 4]
```diff
@@ -0,0 +1,11 @@
+Feature: Handle a well formed packet coming from an agent
```
Comment (zeulus) [inline hub/tests/features/HandleWellFormedPacket.feature]: in below stories there's no distinguishing between "packet" as json data coming from agent, and "packet" as an object of Packet class - they shouldn't be called with the same name

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Feature: Handle a wrong formed packed
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: "wrongly formed" = "malformed"

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  Scenario: Auth info fails to validate because authorization info
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: It's not English and it doesn't say anything: "potato fails to validate because potato" <-- ???

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    Given A packet coming from an agent with the following "metadata" wrong:
+      | auth | invalidAuthKey |
+    When The packet is prepared to be verified against a schema
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: Again, schema validation and auth info validation and 'source' plugin validation are three different validations, schema validation doesn't validate auth info - auth info is validated by during packet validation as one of three phases - schema validation, auth info validation and source validation

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    Then The plugin service fails to validate
+     And The packet is rejected with reason "the authinfo is invalid"
+
+  Scenario: Incoming packet cannot be handled by the system because invalid source
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: because **of** invalid source

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    Given A packet coming from an agent with the following "metadata" wrong:
+      | source | nonRegisteredPlugin |
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: no**t**RegisteredPlugin

File: hub/tests/features/HandleWrongFormedPacket.feature [hunk 10]
```diff
@@ -0,0 +1,21 @@
+Feature: Handle a wrong formed packed
+
+  Scenario: Auth info fails to validate because authorization info
+    Given A packet coming from an agent with the following "metadata" wrong:
+      | auth | invalidAuthKey |
+    When The packet is prepared to be verified against a schema
+    Then The plugin service fails to validate
```
Comment (zeulus) [inline hub/tests/features/HandleWrongFormedPacket.feature]: It's not the plugin service that fails to validate, it's the data that's invalid. That should be rephrased.

File: hub/tests/features/README.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+Builds a well formed agent packet fixture.
+
+#### Wrong formed packet
+```
+@Given /^A packet coming from an agent with the following "([^"]*)" wrong:$/
+```
+
+Builds a wrong formed agent packet with option to specify a `key` where wrong fields will be applied. A `TableNode` object is used to apply the desired wrong fields.
```
Comment (zeulus) [inline hub/tests/features/README.md]: wrong formed -> wrongly formed or just malformed

File: hub/tests/features/README.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+Note that this step **requires** a previously well formed packet. Otherwise the step will be marked as failed.
+
+#### Create a domain packet
+
+```
+@Then /^The system can create a packet in the domain$/
+```
+
+It creates a Domain Packet with using data provided from an Agent Fixture Provider
```
Comment (zeulus) [inline hub/tests/features/README.md]: "using data provided" or "with data provided" - "with using" is duplication

File: hub/tests/features/README.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+Note that this step **requires** a previously well||wrong formed packet. Otherwise the step will be marked as failed.
+
+#### Plugin service validation successes
+
+```
+@Then /^The plugin service success to validate$/
+```
+
+Checks if the plugin service validation has been completed successfully
```
Comment (zeulus) [inline hub/tests/features/README.md]: what is plugin validation service ? something that validates plugins?

File: hub/tests/features/README.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Plugin service validation fails
+
+```
+@Then /^The plugin service fails to validate$/
```
Comment (zeulus) [inline hub/tests/features/README.md]: same as above

File: hub/tests/features/README.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+Checks if the plugin service validation has been failed
+
+#### Packet is rejected with a particular reason
+
+```
+@Given /^The packet is rejected with reason "([^"]*)"$/
+```
+
+Checks the failed validation reasons from Plugin Validation Services.
```
Comment (zeulus) [inline hub/tests/features/README.md]: reason of failure, or fail message, or reason why validation failed

File: hub/tests/features/bootstrap/AgentPacketContext.php [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        self::$agentPacketFixtureProvider = new AgentPacketFixtureProvider;
+    }
+
+    /**
+     * @Given /^A packet coming from an agent with invalid structure$/
+     */
+    public function aPacketComingFromAnAgentWithInvalidStructure()
+    {
+        $this->agentPacket = self::$agentPacketFixtureProvider
+            ->wrongStructuredPacket()
```
Comment (zeulus) [inline hub/tests/features/bootstrap/AgentPacketContext.php]: ->malformedPacket()

File: hub/tests/features/bootstrap/AgentPacketContext.php [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            ->build();
+    }
+
+    /**
+     * @Given /^A packet coming from an agent with the following "([^"]*)" wrong:$/
+     */
+    public function aPacketComingFromAnAgentWithTheFollowingFieldWrong($key, TableNode $table)
+    {
+        $this->agentPacket = self::$agentPacketFixtureProvider
+            ->wrongPopulatedPacket($key, $table->getRows())
```
Comment (zeulus) [inline hub/tests/features/bootstrap/AgentPacketContext.php]: wrongPopulatedPacket() -> packetWithWrongData()

File: hub/tests/features/bootstrap/DIC/definitions.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    $pluginsRepository->retrieve($pluginsCollection);
+
+    return new \FP\Larmo\Application\PluginService($pluginsCollection);
+});
+
+$container['authinfo'] = $container->share(function ($container) {
+    return new \FP\Larmo\Infrastructure\Adapter\IniFileAuthInfoProvider($container['config.path.authinfo']);
+});
+
+$container['json_schema_validation'] = function () {
```
Comment (zeulus) [inline hub/tests/features/bootstrap/DIC/definitions.php]: Why it's not shared()?

File: hub/tests/features/bootstrap/DIC/definitions.php [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    return new \FP\Larmo\Application\Adapter\VendorJsonSchemaValidation;
+};
+
+$container['packet_validation.service'] = function ($container) {
```
Comment (zeulus) [inline hub/tests/features/bootstrap/DIC/definitions.php]: same here, why a new instance every time it's requested?

File: hub/api/dic.php (hunk 1 above)
Comment (dalailomo) [inline hub/api/dic.php]: Adrian wanted to do this:
https://github.com/adrianpietka/larmo/blob/master/hub/tests/unit/Infrastructure/FilesystemPluginsRepositoryTest.php

File: hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * Empty this packet and sets it as a wrong structured one
+     *
+     * Used to test behaviour when comparing against a schema
+     *
+     * @return $this
+     */
+    public function wrongStructuredPacket()
```
Comment (zeulus) [inline hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php]: malformed

File: hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * Force specific wrong fields in desired key
+     *
+     * @param $key
+     */
+    private function forceWrongFields($key)
+    {
+        array_walk_recursive($this->agentPacket[$key], function(&$element, $key) {
```
Comment (zeulus) [inline hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php]: voodoo magic, what does it exactly do? KISS !

File: hub/tests/features/bootstrap/PluginContext.php [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+     * @Given /^The packet is rejected with reason "([^"]*)"$/
+     */
+    public function thePacketIsRejectedWithReason($rejectionReason)
+    {
+        foreach (self::$packetValidationServiceInstance->getErrors() as $singleError) {
+            if ($singleError['message'] === $rejectionReason) {
+                break;
+            }
+
+            throw new Exception('Expected rejection reason not found: '.$rejectionReason);
```
Comment (zeulus) [inline hub/tests/features/bootstrap/PluginContext.php]: why on Earth it's in the foreach? the function is just checking if the *first* error message is equal to provided `$rejectionReason` - if there are more error messages then it's not checking second and each next message

File: hub/api/dic.php (hunk 1 above)
Comment (zeulus) [inline hub/api/dic.php]: @adrianpietka - that pluginRepository should be responsible for whatever it needs to create to dig through filesystem. If that plugin uses Symfony filesystem component, or any other filesystem class, it should not be concern of the calling code to create it, just the plugin repo should be aware what's under the hood

Comment (adrianpietka) [inline hub/api/dic.php]: I want pass instance of DirectoryIterator class to FilesystemPlugins. Not replace - it's no sense.

Comment (zeulus) [inline hub/api/dic.php]: @adrianpietka: there is passing, but below - you have to look at full changes diff, not just the one in comments (https://github.com/adrianpietka/larmo/pull/62/files#diff-9a1f3464e295d9e506a480bbd86aa92dR6)

Comment (zeulus) [inline hub/api/dic.php]: Okay, that code is okay then if @adrianpietka wants to separate this.

File: hub/tests/features/bootstrap/DIC/definitions.php (hunk 18 above)
Comment (dalailomo) [inline hub/tests/features/bootstrap/DIC/definitions.php]: Because I decided to make it in the same way as it it's done here: https://github.com/adrianpietka/larmo/blob/master/hub/api/dic.php#L32

Should we make it shared in both places?

File: hub/tests/features/bootstrap/DIC/definitions.php (hunk 19 above)
Comment (dalailomo) [inline hub/tests/features/bootstrap/DIC/definitions.php]: same as above

File: hub/tests/features/bootstrap/DIC/definitions.php (hunk 18 above)
Comment (zeulus) [inline hub/tests/features/bootstrap/DIC/definitions.php]: @adrianpietka  - what do you think? it's your code probably (json validation services)

File: hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php (hunk 21 above)
Comment (dalailomo) [inline hub/tests/features/bootstrap/Fixtures/Providers/AgentPacket.php]: Is the only solution I found to modify a value in a multidimensional array without using nested foreach loops.

I know this piece of code does not follow the KISS principle but it allows us to write steps like:

```
Given An Agent Packet with the following wrong "metadata" fields:
      | auth | invalidAuthKey |
      | source | invalidSource |
```

For example, in this case, it could be used to check if the auth info is validated before the source info.

Other solution is to create fixtures with specific wrong fields for each case. It's less complicated, but also is less dynamic.
